### PR TITLE
Cancel RPC when senders and receivers are dropped

### PR DESCRIPTION
**Pull request by overvenus**
So we will not accidentally leak gRPC calls.

Cc https://github.com/pingcap/grpc-rs/issues/150

### CODE REVIEW

File: src/call/client.rs [hunk 1]
```diff
@@ -310,6 +316,13 @@ impl<P> StreamingCallSink<P> {
     }
 }
 
+impl<P> Drop for StreamingCallSink<P> {
+    /// The corresponding RPC will be canceled when the value is dropped.
+    fn drop(&mut self) {
+        self.cancel();
```
Comment (BusyJay) [inline src/call/client.rs]: Should it check whether `close` is called?

File: src/call/server.rs [hunk 2]
```diff
@@ -367,23 +375,37 @@ macro_rules! impl_stream_sink {
...
 
+        impl<T> Drop for $t<T> {
+            /// The corresponding RPC will be canceled if the sink did not call
+            /// [`close`] or [`fail`] before dropping.
+            ///
+            /// [`close`]: #method.close
+            /// [`fail`]: #method.fail
+            fn drop(&mut self) {
+                // We did not close it explicitly and it was not dropped in the `fail`.
+                if self.flush_f.is_none() && self.call.is_some() {
```
Comment (siddontang) [inline src/call/server.rs]: why check flush here?

Comment (overvenus) [reply src/call/server.rs]: Check if we have closed it.

Comment (siddontang) [reply src/call/server.rs]: got it

File: src/call/client.rs [hunk 3]
```diff
@@ -421,6 +456,14 @@ impl<H: ShareCallHolder, T> ResponseStreamImpl<H, T> {
             }
         }
     }
+
+    // Cancel the call if we still have some messages or did not
+    // receive status code.
+    fn on_drop(&mut self) {
+        if !self.read_done || !self.finished {
```
Comment (BusyJay) [inline src/call/client.rs]: When will `finished` be true while `read_done` be false?

File: src/call/mod.rs [hunk 4]
```diff
@@ -483,6 +483,14 @@ impl StreamingBase {
             Ok(Async::Ready(bytes))
         }
     }
+
+    // Cancel the call if we still have some messages or did not
+    // receive status code.
+    fn on_drop<C: ShareCallHolder>(&self, call: &mut C) {
+        if !self.read_done || self.close_f.is_some() {
```
Comment (BusyJay) [inline src/call/mod.rs]: Should it be `is_none`?

**Comment by overvenus** [comment]
PTAL

File: src/call/client.rs (hunk 3 above)
Comment (overvenus) [reply src/call/client.rs]: When there are some messages buffered in the memory.

File: src/call/mod.rs (hunk 4 above)
Comment (overvenus) [reply src/call/mod.rs]: I think it should be `is_some`. `Some` means the call is alive.

File: src/call/server.rs (hunk 2 above)
Comment (BusyJay) [reply src/call/server.rs]: I think you can set a flag explicitly at L447.

File: tests/cases/cancel.rs [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        *record_route_handler = Some(Box::new(|stream, sink| {
+            // Start the call and keep the stream and the sink.
+            let f = stream.for_each(|_| Ok(())).then(|_| {
+                let _sink = sink;
+                Ok(())
+            });
+            Box::new(f)
+        }));
+    }
+    {
```
Comment (BusyJay) [inline tests/cases/cancel.rs]: Why brackets?

File: tests/cases/cancel.rs [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+                .join(future::result(Ok(())).map(move |_| {
+                    drop($sink);
+                }))
+                .then(|_| Ok(()));
+            Box::new(f)
+        }}
+    }
+
+    // Start the call, drop the stream and keep the sink.
+    macro_rules! drop_stream {
```
Comment (BusyJay) [inline tests/cases/cancel.rs]: Why not use function?

**Comment by overvenus** [comment]
PTAL, thanks!

File: tests/cases/cancel.rs [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    match rx.into_future().wait() {
+        Err((Error::RpcFailure(s), _)) | Err((Error::RpcFinished(Some(s)), _)) => {
+            assert_eq!(s.status, RpcStatusCode::Cancelled)
+        }
+        Err((e, _)) => panic!("expected cancel, but got: {:?}", e),
+        Ok(_) => panic!("expected error, but got: Ok(_)"),
+    }
+}
+
+fn run_suite() -> (CancelService, RouteGuideClient, Server) {
```
Comment (Hoverbear) [inline tests/cases/cancel.rs]: This name is a bit misleading. `run_suite` reads like it runs some given test suite, but instead it builds some common structures.

File: src/call/client.rs [hunk 8]
```diff
@@ -383,17 +432,19 @@ impl<H: ShareCallHolder, T> ResponseStreamImpl<H, T> {
     }
 
     fn poll(&mut self) -> Poll<Option<T>, Error> {
-        let mut finished = false;
-        self.call.call(|c| {
-            if c.finished {
-                finished = true;
-                return Ok(());
-            }
+        {
```
Comment (BusyJay) [inline src/call/client.rs]: How about check `if !self.finished` first?

Comment (BusyJay) [reply src/call/client.rs]: Any updates?

**Pull request merged**